File: src/app/utils/weatherApi.ts
import type { WeatherData, ForecastData, AirQuality } from '../types/weather';

const API_KEY = '7e29173a70b89b4919dfe873c2352b30';
const BASE_URL = 'https://api.openweathermap.org/data/2.5';

interface GeoResult {
  lat: number;
  lon: number;
  name: string;
  country: string;
}

/**
 * Open-Meteo 지오코딩 API로 도시명(한글·영문 모두 지원)을 위경도로 변환
 */
async function geocodeCity(input: string): Promise<GeoResult> {
  // 영문 + 국가코드 형식(예: Seoul,KR)이면 도시명만 추출
  const cityName = input.split(',')[0].trim();

  const res = await fetch(
    `https://geocoding-api.open-meteo.com/v1/search?name=${encodeURIComponent(cityName)}&count=1&language=ko&format=json`
  );
  if (!res.ok) throw new Error('위치를 찾을 수 없습니다');

  const data = await res.json();
  if (!data.results?.length) throw new Error('도시를 찾을 수 없습니다. 다른 도시명으로 검색해 보세요.');

  const { latitude, longitude, name, country_code } = data.results[0];
  return { lat: latitude, lon: longitude, name, country: country_code };
}

async function reverseGeocodeKo(lat: number, lon: number): Promise<string | undefined> {
  try {
    const res = await fetch(
      `https://nominatim.openstreetmap.org/reverse?lat=${lat}&lon=${lon}&format=json&accept-language=ko`,
      { headers: { 'User-Agent': 'WeatherCodiApp/1.0' } }
    );
    if (!res.ok) return undefined;
    const data = await res.json();
    const addr = data.address ?? {};
    const parts = [
      addr.city_district ?? addr.county,
      addr.suburb ?? addr.neighbourhood ?? addr.quarter,
    ].filter(Boolean);
    return parts.length ? parts.join(' ') : undefined;
  } catch {
    return undefined;
  }
}

export async function fetchWeatherByCoords(lat: number, lon: number): Promise<WeatherData> {
  const [weatherRes, district] = await Promise.all([
    fetch(`${BASE_URL}/weather?lat=${lat}&lon=${lon}&appid=${API_KEY}&units=metric&lang=kr`),
    reverseGeocodeKo(lat, lon),
  ]);
  if (!weatherRes.ok) throw new Error('날씨 정보를 가져올 수 없습니다');
  const d = await weatherRes.json();
  return {
    city: d.name,
    country: d.sys.country === 'KR' ? '대한민국' : d.sys.country,
    district,
    temperature: d.main.temp,
    feelsLike: d.main.feels_like,
    description: d.weather[0].description,
    condition: d.weather[0].main,
    humidity: d.main.humidity,
    windSpeed: d.wind.speed,
    pressure: d.main.pressure,
    visibility: d.visibility,
    sunrise: d.sys.sunrise,
    sunset: d.sys.sunset,
  };
}

export async function fetchWeatherData(city: string): Promise<WeatherData> {
  const { lat, lon, name: koName } = await geocodeCity(city);

  const [res, district] = await Promise.all([
    fetch(`${BASE_URL}/weather?lat=${lat}&lon=${lon}&appid=${API_KEY}&units=metric&lang=kr`),
    reverseGeocodeKo(lat, lon),
  ]);
  if (!res.ok) throw new Error('날씨 정보를 가져올 수 없습니다');

  const d = await res.json();

  if (d.sys.country !== 'KR') {
    throw new Error('국내 도시만 검색할 수 있습니다');
  }

  return {
    city: koName || d.name,
    country: '대한민국',
    district,
    temperature: d.main.temp,
    feelsLike: d.main.feels_like,
    description: d.weather[0].description,
    condition: d.weather[0].main,
    humidity: d.main.humidity,
    windSpeed: d.wind.speed,
    pressure: d.main.pressure,
    visibility: d.visibility,
    sunrise: d.sys.sunrise,
    sunset: d.sys.sunset,
  };
}

// WMO 날씨 코드 → condition/description 매핑
const WMO_MAP: Record<number, { condition: string; description: string }> = {
  0:  { condition: 'Clear',        description: '맑음' },
  1:  { condition: 'Clear',        description: '대체로 맑음' },
  2:  { condition: 'Clouds',       description: '구름 조금' },
  3:  { condition: 'Clouds',       description: '흐림' },
  45: { condition: 'Fog',          description: '안개' },
  48: { condition: 'Fog',          description: '안개' },
  51: { condition: 'Drizzle',      description: '가벼운 이슬비' },
  53: { condition: 'Drizzle',      description: '이슬비' },
  55: { condition: 'Drizzle',      description: '강한 이슬비' },
  61: { condition: 'Rain',         description: '가벼운 비' },
  63: { condition: 'Rain',         description: '비' },
  65: { condition: 'Rain',         description: '강한 비' },
  71: { condition: 'Snow',         description: '가벼운 눈' },
  73: { condition: 'Snow',         description: '눈' },
  75: { condition: 'Snow',         description: '강한 눈' },
  77: { condition: 'Snow',         description: '싸락눈' },
  80: { condition: 'Rain',         description: '소나기' },
  81: { condition: 'Rain',         description: '소나기' },
  82: { condition: 'Rain',         description: '강한 소나기' },
  85: { condition: 'Snow',         description: '눈 소나기' },
  86: { condition: 'Snow',         description: '강한 눈 소나기' },
  95: { condition: 'Thunderstorm', description: '천둥번개' },
  96: { condition: 'Thunderstorm', description: '우박 동반 천둥' },
  99: { condition: 'Thunderstorm', description: '강한 우박 동반 천둥' },
};

function wmoLookup(code: number): { condition: string; description: string } {
  return WMO_MAP[code] ?? { condition: 'Clouds', description: '흐림' };
}

export interface OpenMeteoResult {
  forecast: ForecastData[];
  todayUvIndex: number;
  todayPrecipProb: number;
  airQuality: AirQuality | null;
}

export async function fetchOpenMeteoByCoords(lat: number, lon: number): Promise<OpenMeteoResult> {
  return fetchOpenMeteoForecast(lat, lon);
}

async function fetchOpenMeteoForecast(latitude: number, longitude: number): Promise<OpenMeteoResult> {
  const [fRes, aqRes] = await Promise.allSettled([
    fetch(
      `https://api.open-meteo.com/v1/forecast?latitude=${latitude}&longitude=${longitude}` +
      `&daily=temperature_2m_max,temperature_2m_min,weathercode,precipitation_probability_max,uv_index_max` +
      `&timezone=auto&forecast_days=15`
    ),
    fetch(
      `https://air-quality-api.open-meteo.com/v1/air-quality?latitude=${latitude}&longitude=${longitude}` +
      `&hourly=pm10,pm2_5,european_aqi&timezone=auto&forecast_days=1`
    ),
  ]);

  if (fRes.status === 'rejected' || !fRes.value.ok) throw new Error('예보를 가져올 수 없습니다');
  const fData = await fRes.value.json();

  const daily = fData.daily;
  const forecast: ForecastData[] = (daily.time as string[]).map((dateStr, i) => {
    const date = new Date(dateStr + 'T12:00:00');
    const { condition, description } = wmoLookup(daily.weathercode[i]);
    const tempMax: number = daily.temperature_2m_max[i];
    const tempMin: number = daily.temperature_2m_min[i];
    return {
      date: date.toLocaleDateString('ko-KR', { weekday: 'short', month: 'numeric', day: 'numeric' }),
      temp: Math.round((tempMax + tempMin) / 2),
      tempMin,
      tempMax,
      condition,
      description,
      precipitationProbability: daily.precipitation_probability_max[i] ?? 0,
    };
  });

  let airQuality: AirQuality | null = null;
  if (aqRes.status === 'fulfilled' && aqRes.value.ok) {
    try {
      const aqData = await aqRes.value.json();
      const times = aqData.hourly?.time as string[] | undefined;
      if (times?.length) {
        const nowHour = new Date().toISOString().slice(0, 13);
        let idx = times.findIndex(t => t.slice(0, 13) >= nowHour);
        if (idx < 0) idx = times.length - 1;
        airQuality = {
          pm25: Math.round(aqData.hourly.pm2_5[idx] ?? 0),
          pm10: Math.round(aqData.hourly.pm10[idx] ?? 0),
          aqi: Math.round(aqData.hourly.european_aqi[idx] ?? 0),
        };
      }
    } catch { /* AQI 실패 시 null 유지 */ }
  }

  return {
    forecast,
    todayUvIndex: daily.uv_index_max[0] ?? 0,
    todayPrecipProb: daily.precipitation_probability_max[0] ?? 0,
    airQuality,
  };
}

export async function fetchOpenMeteoData(city: string): Promise<OpenMeteoResult> {
  const { lat, lon } = await geocodeCity(city);
  return fetchOpenMeteoForecast(lat, lon);
}

File: lib/services/weather_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/weather_data.dart';

const _owmKey = '7e29173a70b89b4919dfe873c2352b30';
const _owmBase = 'https://api.openweathermap.org/data/2.5';

const _wmoMap = <int, (String, String)>{
  0:  ('Clear',        '맑음'),
  1:  ('Clear',        '대체로 맑음'),
  2:  ('Clouds',       '구름 조금'),
  3:  ('Clouds',       '흐림'),
  45: ('Fog',          '안개'),
  48: ('Fog',          '안개'),
  51: ('Drizzle',      '가벼운 이슬비'),
  53: ('Drizzle',      '이슬비'),
  55: ('Drizzle',      '강한 이슬비'),
  61: ('Rain',         '가벼운 비'),
  63: ('Rain',         '비'),
  65: ('Rain',         '강한 비'),
  71: ('Snow',         '가벼운 눈'),
  73: ('Snow',         '눈'),
  75: ('Snow',         '강한 눈'),
  77: ('Snow',         '싸락눈'),
  80: ('Rain',         '소나기'),
  81: ('Rain',         '소나기'),
  82: ('Rain',         '강한 소나기'),
  85: ('Snow',         '눈 소나기'),
  86: ('Snow',         '강한 눈 소나기'),
  95: ('Thunderstorm', '천둥번개'),
  96: ('Thunderstorm', '우박 동반 천둥'),
  99: ('Thunderstorm', '강한 우박 동반 천둥'),
};

(String, String) _wmoLookup(int code) =>
    _wmoMap[code] ?? ('Clouds', '흐림');

Future<({double lat, double lon, String name, String country})> geocodeCity(String input) async {
  final city = input.split(',').first.trim();
  final uri = Uri.parse(
    'https://geocoding-api.open-meteo.com/v1/search'
    '?name=${Uri.encodeComponent(city)}&count=1&language=ko&format=json',
  );
  final res = await http.get(uri);
  if (res.statusCode != 200) throw Exception('위치를 찾을 수 없습니다');
  final data = jsonDecode(res.body) as Map<String, dynamic>;
  final results = data['results'] as List?;
  if (results == null || results.isEmpty) {
    throw Exception('도시를 찾을 수 없습니다. 다른 도시명으로 검색해 보세요.');
  }
  final r = results.first as Map<String, dynamic>;
  return (
    lat: (r['latitude'] as num).toDouble(),
    lon: (r['longitude'] as num).toDouble(),
    name: r['name'] as String,
    country: r['country_code'] as String,
  );
}

Future<String?> reverseGeocodeKo(double lat, double lon) async {
  try {
    final uri = Uri.parse(
      'https://nominatim.openstreetmap.org/reverse'
      '?lat=$lat&lon=$lon&format=json&accept-language=ko',
    );
    final res = await http.get(uri, headers: {'User-Agent': 'WeatherCodiApp/1.0'});
    if (res.statusCode != 200) return null;
    final data = jsonDecode(res.body) as Map<String, dynamic>;
    final addr = (data['address'] as Map<String, dynamic>?) ?? {};
    final parts = <String>[
      if (addr['city_district'] != null) addr['city_district'] as String
      else if (addr['county'] != null) addr['county'] as String,
      if (addr['suburb'] != null) addr['suburb'] as String
      else if (addr['neighbourhood'] != null) addr['neighbourhood'] as String
      else if (addr['quarter'] != null) addr['quarter'] as String,
    ].where((s) => s.isNotEmpty).toList();
    return parts.isNotEmpty ? parts.join(' ') : null;
  } catch (_) {
    return null;
  }
}

Future<WeatherData> fetchWeatherByCoords(double lat, double lon) async {
  final results = await Future.wait([
    http.get(Uri.parse(
      '$_owmBase/weather?lat=$lat&lon=$lon&appid=$_owmKey&units=metric&lang=kr',
    )),
    reverseGeocodeKo(lat, lon),
  ]);
  final weatherRes = results[0] as http.Response;
  final district = results[1] as String?;
  if (weatherRes.statusCode != 200) throw Exception('날씨 정보를 가져올 수 없습니다');
  final d = jsonDecode(weatherRes.body) as Map<String, dynamic>;
  final main = d['main'] as Map<String, dynamic>;
  final wind = d['wind'] as Map<String, dynamic>;
  final weather = (d['weather'] as List).first as Map<String, dynamic>;
  final sys = d['sys'] as Map<String, dynamic>;
  return WeatherData(
    city: d['name'] as String,
    country: sys['country'] == 'KR' ? '대한민국' : sys['country'] as String,
    district: district,
    temperature: (main['temp'] as num).toDouble(),
    feelsLike: (main['feels_like'] as num).toDouble(),
    description: weather['description'] as String,
    condition: weather['main'] as String,
    humidity: main['humidity'] as int,
    windSpeed: (wind['speed'] as num).toDouble(),
    pressure: main['pressure'] as int,
    visibility: d['visibility'] as int,
  );
}

Future<WeatherData> fetchWeatherData(String city) async {
  final geo = await geocodeCity(city);
  final results = await Future.wait([
    http.get(Uri.parse(
      '$_owmBase/weather?lat=${geo.lat}&lon=${geo.lon}&appid=$_owmKey&units=metric&lang=kr',
    )),
    reverseGeocodeKo(geo.lat, geo.lon),
  ]);
  final weatherRes = results[0] as http.Response;
  final district = results[1] as String?;
  if (weatherRes.statusCode != 200) throw Exception('날씨 정보를 가져올 수 없습니다');
  final d = jsonDecode(weatherRes.body) as Map<String, dynamic>;
  final sys = d['sys'] as Map<String, dynamic>;
  if (sys['country'] != 'KR') throw Exception('국내 도시만 검색할 수 있습니다');
  final main = d['main'] as Map<String, dynamic>;
  final wind = d['wind'] as Map<String, dynamic>;
  final weather = (d['weather'] as List).first as Map<String, dynamic>;
  return WeatherData(
    city: geo.name.isNotEmpty ? geo.name : d['name'] as String,
    country: '대한민국',
    district: district,
    temperature: (main['temp'] as num).toDouble(),
    feelsLike: (main['feels_like'] as num).toDouble(),
    description: weather['description'] as String,
    condition: weather['main'] as String,
    humidity: main['humidity'] as int,
    windSpeed: (wind['speed'] as num).toDouble(),
    pressure: main['pressure'] as int,
    visibility: d['visibility'] as int,
  );
}

Future<OpenMeteoResult> fetchOpenMeteoByCoords(double lat, double lon) async {
  return _fetchOpenMeteoForecast(lat, lon);
}

Future<OpenMeteoResult> fetchOpenMeteoData(String city) async {
  final geo = await geocodeCity(city);
  return _fetchOpenMeteoForecast(geo.lat, geo.lon);
}

Future<OpenMeteoResult> _fetchOpenMeteoForecast(double lat, double lon) async {
  final uri = Uri.parse(
    'https://api.open-meteo.com/v1/forecast'
    '?latitude=$lat&longitude=$lon'
    '&daily=temperature_2m_max,temperature_2m_min,weathercode,precipitation_probability_max,uv_index_max'
    '&timezone=auto&forecast_days=15',
  );
  final res = await http.get(uri);
  if (res.statusCode != 200) throw Exception('예보를 가져올 수 없습니다');
  final data = jsonDecode(res.body) as Map<String, dynamic>;
  final daily = data['daily'] as Map<String, dynamic>;
  final times = daily['time'] as List;
  final forecast = List<ForecastData>.generate(times.length, (i) {
    final date = DateTime.parse('${times[i]}T12:00:00');
    final wday = ['일', '월', '화', '수', '목', '금', '토'][date.weekday % 7];
    final dateStr = '${date.month}월 ${date.day}일($wday)';
    final (condition, description) = _wmoLookup(daily['weathercode'][i] as int);
    final tempMax = (daily['temperature_2m_max'][i] as num).toDouble();
    final tempMin = (daily['temperature_2m_min'][i] as num).toDouble();
    return ForecastData(
      date: dateStr,
      temp: ((tempMax + tempMin) / 2).roundToDouble(),
      tempMin: tempMin,
      tempMax: tempMax,
      condition: condition,
      description: description,
      precipitationProbability: (daily['precipitation_probability_max'][i] as num?)?.toInt() ?? 0,
    );
  });
  return OpenMeteoResult(
    forecast: forecast,
    todayUvIndex: (daily['uv_index_max'][0] as num?)?.toDouble() ?? 0,
    todayPrecipProb: (daily['precipitation_probability_max'][0] as num?)?.toInt() ?? 0,
  );
}